calling-apis: content (#15)
Ada-Developers-Academy/core-unit-3: calling-apis/problem-set-requests-with-js.md

# Problem Set: Requests With JS

## Directions

Read the problem prompt, which will ask you to write JavaScript code to find certain answers.

Once you found the answers, input them into the answers prompt below.

## Preparation: LocationIQ Forward Geocoding API

Use the [LocationIQ's Forward Geocoding API](https://locationiq.com/geocoding). Before starting this problem prompt:

1. Obtain an API key for LocationIQ's Forward Geocoding API
1. Familiarize yourself with the LocationIQ API Documentation
1. Familiarize yourself with the `axios` library
1. Open Postman for manual testing

![Screenshot of LocationIQ's feature on the Geocoding API](../assets/calling-apis_problem-set-requests-with-js_location-iq-geocoding.png)

## Problem Prompt

Create a JavaScript program that finds the latitude and longitude coordinates for the [seven wonders of the New7Wonders of the World](https://en.wikipedia.org/wiki/New7Wonders_of_the_World), which are the following:

| Wonder              |
| ------------------- |
| Great Wall of China |

> | Petra               |

Reviewer: I've frequently getting back an Austrian hairdresser shop called Petra when the query is just "Petra". Adding in Jordan helps avoid this from happening, but other locations also give strange results some times, so I'm not sure this is worth changing...

Author: Seems like we'll have to rethink autograding on Seven Wonders if it's inconsistent...

| Colosseum           |
| Chichen Itza        |
| Machu Picchu        |
| Taj Mahal           |
| Christ the Redeemer |

> 

Reviewer: provide a suggestion about how to approach this. I'm extrapolating from your comments in the instructor notes and explicitly adding the recommendation that students start simply with a single request, with the rate limit reminder up here so they don't accidentally lock their key if they get started without scrolling all the way down to the current warning.
```suggestion

Remember that many API providers will deactivate an API key if they detect too many requests within a short span of time. Since adding pauses or timeouts to JavaScript code can be tricky, it may be worth starting with a program that retrieves the data for a single location, and attempt the Extra Challenges after that is working.

```

### Extra Challenges

> See if you can accomplish these extra challenges. These challenges are tricky and interesting! One solution required extra research around the following subjects: `setTimeout`, "JavaScript sleep," `async` and `await`, and (traditional) `for` loops.

Reviewer: Just say that researching these topics may be required rather than referencing some solution they can't see.
```suggestion
See if you can accomplish these extra challenges. These challenges are tricky and interesting! Implementing these extensions may require extra research around the following subjects: `setTimeout`, "JavaScript sleep," `async` and `await`, and (traditional) `for` loops.
```

1. Organize your code into functions and helper functions
1. Use a loop
1. Create this data structure to hold your answers

```js
{
    "Great Wall of China": {
        latitude: "...",
        longitude: "..."
    },
    "Petra": {
        latitude: "...",
        longitude: "..."
    },
    "Colosseum": {
        latitude: "...",
        longitude: "..."
    },
    "Chichen Itza": {
        latitude: "...",
        longitude: "..."
    },
    "Machu Picchu": {
        latitude: "...",
        longitude: "..."
    },
    "Taj Mahal": {
        latitude: "...",
        longitude: "..."
    },
    "Christ the Redeemer": {
        latitude: "...",
        longitude: "..."
    }
}
```

### !callout-warning

## Hint: Working with Rate Limits

Rapidly making several HTTP requests to the same endpoint may exceed our API rate limit. There are many ways to work with the rate limit. As a suggestion, one strategy is to consider the following:

1. How can we make requests more slowly?

> 1. Research the `setTimeout` function, or how to implement a "sleep" function. We recommend making the program wait for 500 milliseconds to begin your experiment.

Reviewer: rephrase attempting the pause logic so that it could be done in a separate program before bringing their findings into this solution
explicitly state that this is a good rate limiting delay

```suggestion
1. Research the `setTimeout` function, or how to implement a "sleep" function. We recommend trying to make a program wait for 500 milliseconds as a starting point. This is also a good pause to use between API calls!
```

<!-- Question 2 -->
<!-- prettier-ignore-start -->
### !challenge
* type: short-answer
* id: a21964ba
* title: Requests with JS
##### !question
What is the longitude of the first result of "Great Wall of China"?
##### !end-question
##### !answer
116.0136394
##### !end-answer
### !end-challenge
<!-- prettier-ignore-end -->

<!-- Question 3 -->
<!-- prettier-ignore-start -->
### !challenge
* type: short-answer
* id: 5d5b6f83
* title: Requests with JS
##### !question
What is the latitude of the first result of "Petra"?
##### !end-question
##### !answer
30.3273975
##### !end-answer
### !end-challenge
<!-- prettier-ignore-end -->

<!-- Question 4 -->
<!-- prettier-ignore-start -->
### !challenge
* type: short-answer
* id: 7bd1170a
* title: Requests with JS
##### !question
What is the longitude of the first result of "Petra"?
##### !end-question
##### !answer
35.4464006
##### !end-answer
### !end-challenge
<!-- prettier-ignore-end -->

<!-- Question 5 -->
<!-- prettier-ignore-start -->
### !challenge
* type: short-answer
* id: e0a8740f
* title: Requests with JS
##### !question
What is the latitude of the first result of "Colosseum"?
##### !end-question
##### !answer

> 41.8910091

Reviewer: i get a different value for the Colosseum
```suggestion
41.8902614
```

##### !end-answer
### !end-challenge
<!-- prettier-ignore-end -->

<!-- Question 6 -->
<!-- prettier-ignore-start -->
### !challenge
* type: short-answer
* id: 7f475b9a
* title: Requests with JS
##### !question
What is the longitude of the first result of "Colosseum"?
##### !end-question
##### !answer

> 12.4920748

Reviewer: I get a different value for the Colosseum
```suggestion
12.4930871035955
```
